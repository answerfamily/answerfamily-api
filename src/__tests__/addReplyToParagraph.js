jest.mock('../dataloaders/auth0UserLoaderFactory');
jest.mock('../lib/scrapUrls');

const { ObjectId } = require('mongodb');
const {
  loadESFixtures,
  unloadESFixtures,
  loadMongoFixtures,
  unloadMongoFixtures,
} = require('../../test/fixtures');
const gql = require('../../test/gql');
const {
  ES_FIXTURES,
  MONGO_FIXTURES,
} = require('../__fixtures__/articleParagraphReply');

const esClient = require('../lib/esClient');
const mongoClient = require('../lib/mongoClient');
const scrapUrls = require('../lib/scrapUrls');

describe('addReplyToParagraph', () => {
  beforeAll(async () => {
    await loadESFixtures(ES_FIXTURES);
    await loadMongoFixtures(MONGO_FIXTURES);
  });

  afterAll(async () => {
    await unloadESFixtures(ES_FIXTURES);
    await unloadMongoFixtures(MONGO_FIXTURES);
  });

  it('creates reply and paragraphReplies', async () => {
    scrapUrls.mockImplementation(() => Promise.resolve({}));
    const reply = {
      text: '明年五月台灣將成為亞洲第一個同性可合法結婚的國家',
      note: 'notenote',
    };

    const { errors, data } = await gql`
      mutation($paragraphId: String!, $reply: ReplyInput!) {
        addReplyToParagraph(paragraphId: $paragraphId, reply: $reply) {
          paragraphReplies {
            id
            reply {
              text
              note
              user {
                id
              }
            }
            user {
              id
            }
          }
        }
      }
    `(
      {
        paragraphId: 'p1',
        reply,
      },
      {
        userPromise: Promise.resolve({ iss: 'user-id' }),
      }
    );

    expect(errors).toBeUndefined();
    expect(
      // ID is random, thus exclude from snapshot
      // eslint-disable-next-line no-unused-vars
      data.addReplyToParagraph.paragraphReplies.map(({ id, ...data }) => data)
    ).toMatchSnapshot();
    expect(scrapUrls.mock.calls).toMatchSnapshot('scrapUrl calls');

    const newParagraphReply = data.addReplyToParagraph.paragraphReplies.find(
      p => p.reply.text === reply.text
    );

    // cleanup
    const { db } = await mongoClient;
    await esClient.deleteByQuery({
      index: 'replies',
      type: '_doc',
      body: {
        query: {
          term: {
            text: reply.text,
          },
        },
      },
    });
    await db
      .collection('paragraphReplies')
      .deleteOne({ _id: ObjectId(newParagraphReply.id) });
  });

  it('blocks non-logged in users', async () => {
    const { errors } = await gql`
      mutation($paragraphId: String!, $reply: ReplyInput!) {
        addReplyToParagraph(paragraphId: $paragraphId, reply: $reply) {
          id
        }
      }
    `({
      paragraphId: 'p1',
      reply: { text: 'foo' },
    });

    // Should be unauthorzied
    expect(errors).toMatchSnapshot();
  });
});
