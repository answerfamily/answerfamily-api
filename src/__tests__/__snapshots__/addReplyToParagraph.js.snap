// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`addReplyToParagraph blocks non-logged in users 1`] = `
Array [
  [GraphQLError: Must log in first],
]
`;

exports[`addReplyToParagraph creates reply and paragraphReplies 1`] = `
Array [
  Object {
    "reply": Object {
      "note": "notenote",
      "text": "明年五月台灣將成為亞洲第一個同性可合法結婚的國家",
      "user": Object {
        "id": "user-id",
      },
    },
    "user": Object {
      "id": "user-id",
    },
  },
  Object {
    "reply": Object {
      "note": null,
      "text": "「外國同性戀藉結婚來台治療愛滋病，都是健保支付，全民買單」這句話其實並不算錯，但實務上並不會有「拖垮健保」的狀況，也跟同性婚姻關係不大，所以不用太擔心唷。",
      "user": Object {
        "id": "author1",
      },
    },
    "user": Object {
      "id": "author1",
    },
  },
]
`;
